src/repr/src/stats.rs

```diff
@@ -84,123 +83,100 @@ pub(crate) fn jsonb_stats_nulls(
+    let arena = RowArena::new();
     for idx in 0..col.len() {
         let val = ColumnGet::<Option<Vec<u8>>>::get(col, idx);
         if !validity.get(idx) {
             assert!(val.map_or(true, |x| x.is_empty()));
             continue;
         }
-        NullableProtoDatumToPersist::decode(val, &mut buf.packer());
-        let datum = as_optional_datum(&buf).expect("not enough datums");
+        let datum = arena.make_datum(|r| NullableProtoDatumToPersist::decode(val, r));
         // Datum::Null only shows up at the top level of Jsonb, so we handle it
         // here instead of in the recursing function.
         if let Datum::Null = datum {
             null_count += 1;
         } else {
-            let () = jsonb_stats_datum(&mut stats, datum)?;
+            let () = datums.push(datum);
         }
     }
-    Ok((stats, null_count))
+    Ok((datums.to_stats(), null_count))
 }
 
-fn jsonb_stats_datum(stats: &mut JsonStats, datum: Datum<'_>) -> Result<(), String> {
-    fn update_stats<T: PartialOrd + ToOwned + ?Sized>(
-        stats: &mut PrimitiveStats<T::Owned>,
-        val: &T,
-    ) {
-        if val < stats.lower.borrow() {
-            stats.lower = val.to_owned();
-        }
-        if val > stats.upper.borrow() {
-            stats.upper = val.to_owned();
+struct JsonDatums<'a> {
+    count: usize,
+    min_max: Option<(Datum<'a>, Datum<'a>)>,
+    nested: BTreeMap<String, JsonDatums<'a>>,
+}
+
+impl<'a> Default for JsonDatums<'a> {
+    fn default() -> Self {
+        JsonDatums {
+            count: 0,
+            min_max: None,
+            nested: BTreeMap::new(),
         }
     }
+}
 
-    match datum {
-        Datum::JsonNull => match stats {
-            JsonStats::None => *stats = JsonStats::JsonNulls,
-            JsonStats::JsonNulls => {}
-            _ => *stats = JsonStats::Mixed,
-        },
-        Datum::False => match stats {
-            JsonStats::None => {
-                *stats = JsonStats::Bools(PrimitiveStats {
-                    lower: false,
-                    upper: false,
-                })
+impl<'a> JsonDatums<'a> {
+    fn push(&mut self, datum: Datum<'a>) {
+        self.count += 1;
+        self.min_max = match self.min_max.take() {
+            None => Some((datum, datum)),
+            Some((min, max)) => Some((min.min(datum), max.max(datum))),
+        };
+        if let Datum::Map(map) = datum {
+            for (key, val) in map.iter() {
+                let val_datums = self.nested.entry(key.to_owned()).or_default();
+                val_datums.push(val);
             }
-            JsonStats::Bools(stats) => update_stats(stats, &false),
-            _ => *stats = JsonStats::Mixed,
-        },
-        Datum::True => match stats {
-            JsonStats::None => {
-                *stats = JsonStats::Bools(PrimitiveStats {
-                    lower: true,
-                    upper: true,
+        }
+    }
+    fn to_stats(self) -> JsonStats {
+        if self.count == 0 {
```

Comment:
is count used other than here? if not, feels redundant with the `None => JsonStats::None` case in the `match self.min_max`

Reply:
Yep, for the individual field counts on `JsonMapElementStats`. Though this specific check is definitely redundant... removed.